src/gtest/main.cpp

```diff
@@ -18,23 +20,73 @@ struct ECCryptoClosure
 
 ECCryptoClosure instance_of_eccryptoclosure;
 
+class LogGrabber : public ::testing::EmptyTestEventListener {
+    fs::path logPath;
+
+public:
+    LogGrabber(fs::path logPathIn) : logPath(logPathIn) {}
+
+    virtual void OnTestStart(const ::testing::TestInfo& test_info) {
+        // Test logs are written synchronously, so we can clear the log file to
+        // ensure that at the end of the test, the log lines are all related to
+        // the test itself.
+        std::ofstream logFile;
+        logFile.open(logPath.string(), std::ofstream::out | std::ofstream::trunc);
+        logFile.close();
```

Comment:
An open file handle maintains a file position, which [isn't necessarily reset](https://stackoverflow.com/a/984761/393146) to offset 0 of the file when it is truncated. I guess this works because the tracing appender opens the file in `O_APPEND` mode, and that seeks to the end of the file (i.e. offset 0 if it has just been truncated) [before each write](https://pubs.opengroup.org/onlinepubs/9699919799/functions/open.html) on POSIX systems. I would want to test it on Windows and macOS though, since there's a plausible alternative implementation where `O_APPEND` doesn't check for concurrent truncation.

Reply:
The tests aren't using `tracing_appender` (which is non-blocking) after this PR, but instead `Mutex<File>`. Its `MakeWriter` impl is [here](https://github.com/tokio-rs/tracing/blob/8d4d5ac1a95770230046cc860e81d25ae3261a1d/tracing-subscriber/src/fmt/writer.rs#L755-L794), which trivially calls through to the `io::Write` impl on `&mut File`. So the relevant logic is defined by [`File::create`](https://doc.rust-lang.org/stable/std/fs/struct.File.html#method.create) (as called by `tracing_init_test`), which is implemented as:
```rust
OpenOptions::new().write(true).create(true).truncate(true).open(path.as_ref())
```
So the file is not opened AFAICT with `O_APPEND`. However, I don't think it really matters much, because all this is used for is per-test log extraction, and if the append doesn't re-seek on a particular platform then we'll just get a bunch of null bytes printed to the gtest output of failing tests (remember that we don't print anything on success). I've confirmed that it works correctly on Ubuntu 20.04.

Reply:
Sounds good, thanks for the detailed explanation.